packages/multi-select-combo-box/src/vaadin-multi-select-combo-box-internal.js

```diff
@@ -170,20 +170,32 @@ class MultiSelectComboBoxInternal extends ComboBoxDataProviderMixin(ComboBoxMixi
       return;
     }
 
-    if (this.topGroup) {
-      const filteredTopItems = [];
+    if (items && items.length && this.topGroup && this.topGroup.length) {
+      let filteredTopItems = [];
       const filteredItems = [];
+      const topItems = [...this.topGroup];
 
-      (items || []).forEach((item) => {
-        if (this.topGroup.some((selectedItem) => this._getItemValue(item) === this._getItemValue(selectedItem))) {
+      items.forEach((item) => {
+        const idx = topItems.findIndex((selectedItem) => this._getItemValue(item) === this._getItemValue(selectedItem));
+        if (idx !== -1) {
           filteredTopItems.push(item);
+          topItems.splice(idx, 1);
         } else {
           filteredItems.push(item);
         }
       });
 
+      // When there is no filter, always show all the top group,
+      // including items from pages that haven't been loaded yet.
+      if (!this.filter && topItems.length) {
+        filteredTopItems = [...filteredTopItems, ...topItems];
```

Comment:
I'm under the impression that, although the PR fixes the issue for the default case when no filter, it can still occur with filters that return more than one page.

Reply:
When the filter is applied, we don't show the top group in the overlay.

Reply:
Ah, I see. Do we really need to do this heavy logic then when a filter is applied? Could we simply do `this._dropdownItems = items` when there is a filter?

Reply:
Yep, I will update the PR.

Reply:
Actually, current logic is like this: when top group items match the filter, they are still shown on top.
I would prefer not to change this in this PR as it's unrelated to the original issue that it fixes.

Reply:
How does the implementation know which top items match the filter to keep only them? I can't see any logic that would match the top items against the filter.

Reply:
We ended up with not showing top group when filter is applied - covered by #6723